frontend/src/components/question-creation/QuestionElementItem.tsx

```diff
@@ -0,0 +1,145 @@
+import { DragTypes } from "../../types/DragTypes";
+
+export interface QuestionElementItemProps {
+  content: QuestionElement;
+  index: number;
+}
+
+interface DragItem {
+  index: number;
+  id: string;
+  type: string;
+}
+
+const renderQuestionContent = (content: QuestionElement) => {
+  const { id, type, data } = content;
+  switch (type) {
+    case QuestionElementType.QUESTION:
+      return <Text key={id}>this is a question element.</Text>;
+    case QuestionElementType.TEXT:
+      return <TextElement key={id} id={id} data={data} />;
+    case QuestionElementType.IMAGE:
+      return <Text key={id}>this is an image element.</Text>;
+    case QuestionElementType.MULTIPLE_CHOICE:
+      return <Text key={id}>this is a multiple choice element.</Text>;
+    case QuestionElementType.SHORT_ANSWER:
+      return <Text key={id}>this is a short answer element.</Text>;
+    case QuestionElementType.MULTI_SELECT:
+      return <Text key={id}>this is a multi select element.</Text>;
+    default:
+      return null;
+  }
+};
+
+const QuestionElementItem = ({
+  content,
+  index,
+}: QuestionElementItemProps): React.ReactElement => {
+  const { id, error } = content;
+  const { setQuestionElements } = useContext(QuestionEditorContext);
+
+  const reorderQuestionElements = (hoverIndex: number, dragIndex: number) => {
+    setQuestionElements((prevElements) =>
+      update(prevElements, {
+        $splice: [
+          [dragIndex, 1],
+          [hoverIndex, 0, prevElements[dragIndex]],
+        ],
+      }),
+    );
+  };
+
+  const removeQuestionElement = () => {
+    setQuestionElements((prevElements) =>
+      prevElements.filter((element) => element.id !== id),
+    );
+  };
+
+  const dragRef = useRef<HTMLDivElement>(null);
+  const previewRef = useRef<HTMLDivElement>(null);
+  const [{ handlerId }, drop] = useDrop<
+    DragItem,
+    void,
+    { handlerId: Identifier | null }
+  >({
+    accept: DragTypes.QUESTION_ELEMENT_ITEM,
+    collect(monitor) {
+      return {
+        handlerId: monitor.getHandlerId(),
+      };
+    },
+    hover(item: DragItem, monitor) {
+      if (!previewRef.current) {
+        return;
+      }
+      const dragIndex = item.index;
+      const hoverIndex = index;
+      if (dragIndex === hoverIndex) {
+        return;
+      }
+      const hoverBoundingRect = previewRef.current?.getBoundingClientRect();
+      const hoverMiddleY =
+        (hoverBoundingRect.bottom - hoverBoundingRect.top) / 2;
+      const clientOffset = monitor.getClientOffset();
+      const hoverClientY = (clientOffset as XYCoord).y - hoverBoundingRect.top;
+      if (dragIndex < hoverIndex && hoverClientY < hoverMiddleY) {
+        return;
+      }
+      if (dragIndex > hoverIndex && hoverClientY > hoverMiddleY) {
+        return;
+      }
+      reorderQuestionElements(dragIndex, hoverIndex);
+      /* eslint-disable no-param-reassign */
+      item.index = hoverIndex;
```

Comment:
documentation recommends ignoring `no-param-reassign` in this case for performance reasons